CONFIG.md
# Configuration files

The file containing the program configuration parameters to be changed is located in the ./configs directory and is called config.yml. If any property from this file is removed or its value is written incorrectly, it will be replaced with a default value on startup with a warning in the console.

## Structure of config.yml

``` Yaml
db:
  connTimeoutSeconds: 5 // how many seconds it takes to try to connect to the database
  connLatencyMilliseconds: 50 // how many milliseconds will be the delay in sending each request to the database
  host: "127.0.0.1" // what ip is the database server
  logConnStatus: false // whether to display the status of connecting / disconnecting from the database in the logs
  name: "video_info" // name of the database in the DBMS
  port: "5432" // database server port
  sslMode: "disable" // SSL encryption for sending/receiving requests to the database
  username: "postgres" // DBMS username

pagination:
  getLimitDefault: 100 // default number of items to display in get methods

server:
  debugMode: true // set server state to debug mode
  maxHeaderBytes: 1048576 // reserve 1 megabyte for storing request headers per server
  port: "8000" // web server port
  readTimeoutSeconds: 15 // maximum time in seconds to receive responses from the server
  writeTimeoutSeconds: 15 // maximum time in seconds to send requests to the server

session:
  ttlHours: 168 // browser and session cookie lifetime in hours (which is 7 days)
```

# Environment files

For the server application to work, you need to create an .env file in the /configs directory. By default, it is not bundled with the program for security reasons. It contains information for connecting to your database, a "salt" for hashing passwords, and a signing key for generating and reading session tokens. An example of this arbitrary data file is included in the distribution and is called `template.env`.

## Structure of .env

``` Sh
DB_DRIVER=mongodb # name to adjust to your DBMS
DB_PASSWORD='6084213642' # DBMS user password
HASHING_PASSWORD_SALT='aB@#$%1' # salt for passwords hashing
HASHING_TOKEN_SIGNING_KEY='cD@#$%2' # signing key for generating and reading tokens
```


internal/config/config.go
package config

import (
	"os"
	"strconv"
	"strings"
	"sync"

	"github.com/mrumyantsev/video-hosting/internal/logger"
	msg "github.com/mrumyantsev/video-hosting/internal/messages"

	"github.com/spf13/viper"
)

type Config struct {
	DBConnectionLatencyMilliseconds int
	DBConnectionShowStatus          bool
	DBConnectionTimeoutSeconds      int
	DBDriver                        string
	DBHost                          string
	DBName                          string
	DBPort                          int
	DBSSLEnable                     bool
	DBUsername                      string

	DBPassword string

	DBOConnectionLatencyMilliseconds int
	DBOConnectionShowStatus          bool
	DBOConnectionTimeoutSeconds      int
	DBODriver                        string
	DBOHost                          string
	DBOName                          string
	DBOPort                          int
	DBOSSLEnable                     bool
	DBOUsername                      string

	DBOPassword string

	HashingPasswordSalt    string
	HashingTokenSigningKey string

	PaginationGetLimitDefault int

	ServerDebugEnable         bool
	ServerMaxHeaderBytes      int
	ServerHost                string
	ServerPort                int
	ServerReadTimeoutSeconds  int
	ServerWriteTimeoutSeconds int

	SessionTTLHours int

	StreamICEServersMutex            sync.RWMutex
	StreamICEServers                 []string
	StreamLink                       string
	StreamSnapshotPeriodSeconds      int
	StreamSnapshotShowStatus         bool
	StreamSnapshotsEnable            bool
	StreamStreamsUpdatePeriodSeconds int

	ServerIP string
}

func LoadConfig(path string) (*Config, error) {
	// Parse config file path
	path = path[:len(path)-4]
	lastDirIndex := strings.LastIndex(path, "/")
	viper.AddConfigPath(path[:lastDirIndex+1])
	viper.SetConfigName(path[lastDirIndex+1:])

	// Load data from config file
	var cfg Config
	err := viper.ReadInConfig()
	if err != nil {
		return nil, err
	}

	param := "DB_CONNECTION_LATENCY_MILLISECONDS"
	if val, err := strconv.Atoi(os.Getenv(param)); err != nil {
		defaultVal := 50
		cfg.DBConnectionLatencyMilliseconds = defaultVal
		logger.Print(msg.WarningCannotConvertCvar(param, defaultVal))
	} else {
		cfg.DBConnectionLatencyMilliseconds = val
	}

	param = "DB_CONNECTION_SHOW_STATUS"
	if val, err := strconv.ParseBool(os.Getenv(param)); err != nil {
		defaultVal := false
		cfg.DBConnectionShowStatus = defaultVal
		logger.Print(msg.WarningCannotConvertCvar(param, defaultVal))
	} else {
		cfg.DBConnectionShowStatus = val
	}

	param = "DB_CONNECTION_TIMEOUT_SECONDS"
	if val, err := strconv.Atoi(os.Getenv(param)); err != nil {
		defaultVal := 5
		cfg.DBConnectionTimeoutSeconds = defaultVal
		logger.Print(msg.WarningCannotConvertCvar(param, defaultVal))
	} else {
		cfg.DBConnectionTimeoutSeconds = val
	}

	param = "DB_DRIVER"
	if os.Getenv(param) == "" {
		defaultVal := "mysql"
		cfg.DBDriver = defaultVal
		logger.Print(msg.WarningCannotConvertCvar(param, defaultVal))
	} else {
		cfg.DBDriver = os.Getenv(param)
	}

	param = "DB_HOST"
	if os.Getenv(param) == "" {
		defaultVal := "127.0.0.1"
		cfg.DBHost = defaultVal
		logger.Print(msg.WarningCannotConvertCvar(param, defaultVal))
	} else {
		cfg.DBHost = os.Getenv(param)
	}

	param = "DB_NAME"
	if os.Getenv(param) == "" {
		defaultVal := "vhs"
		cfg.DBName = defaultVal
		logger.Print(msg.WarningCannotConvertCvar(param, defaultVal))
	} else {
		cfg.DBName = os.Getenv(param)
	}

	param = "DB_PASSWORD"
	if os.Getenv(param) == "" {
		defaultVal := "1234"
		cfg.DBPassword = defaultVal
		logger.Print(msg.WarningCannotConvertCvar(param, defaultVal))
	} else {
		cfg.DBPassword = os.Getenv(param)
	}

	param = "DB_PORT"
	if val, err := strconv.Atoi(os.Getenv(param)); err != nil {
		defaultVal := 2345
		cfg.DBPort = defaultVal
		logger.Print(msg.WarningCannotConvertCvar(param, defaultVal))
	} else {
		cfg.DBPort = val
	}

	param = "DB_SSL_ENABLE"
	if val, err := strconv.ParseBool(os.Getenv(param)); err != nil {
		defaultVal := false
		cfg.DBSSLEnable = defaultVal
		logger.Print(msg.WarningCannotConvertCvar(param, defaultVal))
	} else {
		cfg.DBSSLEnable = val
	}

	param = "DB_USERNAME"
	if os.Getenv(param) == "" {
		defaultVal := "joe"
		cfg.DBUsername = defaultVal
		logger.Print(msg.WarningCannotConvertCvar(param, defaultVal))
	} else {
		cfg.DBUsername = os.Getenv(param)
	}

	param = "DBO_CONNECTION_LATENCY_MILLISECONDS"
	if val, err := strconv.Atoi(os.Getenv(param)); err != nil {
		defaultVal := 500
		cfg.DBOConnectionLatencyMilliseconds = defaultVal
		logger.Print(msg.WarningCannotConvertCvar(param, defaultVal))
	} else {
		cfg.DBOConnectionLatencyMilliseconds = val
	}

	param = "DBO_CONNECTION_SHOW_STATUS"
	if val, err := strconv.ParseBool(os.Getenv(param)); err != nil {
		defaultVal := false
		cfg.DBOConnectionShowStatus = defaultVal
		logger.Print(msg.WarningCannotConvertCvar(param, defaultVal))
	} else {
		cfg.DBOConnectionShowStatus = val
	}

	param = "DBO_CONNECTION_TIMEOUT_SECONDS"
	if val, err := strconv.Atoi(os.Getenv(param)); err != nil {
		defaultVal := 5
		cfg.DBOConnectionTimeoutSeconds = defaultVal
		logger.Print(msg.WarningCannotConvertCvar(param, defaultVal))
	} else {
		cfg.DBOConnectionTimeoutSeconds = val
	}

	param = "DBO_DRIVER"
	if os.Getenv(param) == "" {
		defaultVal := "mysql"
		cfg.DBODriver = defaultVal
		logger.Print(msg.WarningCannotConvertCvar(param, defaultVal))
	} else {
		cfg.DBODriver = os.Getenv(param)
	}

	param = "DBO_HOST"
	if os.Getenv(param) == "" {
		defaultVal := "127.0.0.1"
		cfg.DBOHost = defaultVal
		logger.Print(msg.WarningCannotConvertCvar(param, defaultVal))
	} else {
		cfg.DBOHost = os.Getenv(param)
	}

	param = "DBO_NAME"
	if os.Getenv(param) == "" {
		defaultVal := "l333"
		cfg.DBOName = defaultVal
		logger.Print(msg.WarningCannotConvertCvar(param, defaultVal))
	} else {
		cfg.DBOName = os.Getenv(param)
	}

	param = "DBO_PASSWORD"
	if os.Getenv(param) == "" {
		defaultVal := "1234"
		cfg.DBOPassword = defaultVal
		logger.Print(msg.WarningCannotConvertCvar(param, defaultVal))
	} else {
		cfg.DBOPassword = os.Getenv(param)
	}

	param = "DBO_PORT"
	if val, err := strconv.Atoi(os.Getenv(param)); err != nil {
		defaultVal := 2345
		cfg.DBOPort = defaultVal
		logger.Print(msg.WarningCannotConvertCvar(param, defaultVal))
	} else {
		cfg.DBOPort = val
	}

	param = "DBO_SSL_ENABLE"
	if val, err := strconv.ParseBool(os.Getenv(param)); err != nil {
		defaultVal := false
		cfg.DBOSSLEnable = defaultVal
		logger.Print(msg.WarningCannotConvertCvar(param, defaultVal))
	} else {
		cfg.DBOSSLEnable = val
	}

	param = "DBO_USERNAME"
	if os.Getenv(param) == "" {
		defaultVal := "john"
		cfg.DBOUsername = defaultVal
		logger.Print(msg.WarningCannotConvertCvar(param, defaultVal))
	} else {
		cfg.DBOUsername = os.Getenv(param)
	}

	param = "HASHING_PASSWORD_SALT"
	if os.Getenv(param) == "" {
		defaultVal := "SdD2Sdf@dFhSe#r"
		cfg.HashingPasswordSalt = defaultVal
		logger.Print(msg.WarningCannotConvertCvar(param, defaultVal))
	} else {
		cfg.HashingPasswordSalt = os.Getenv(param)
	}

	param = "HASHING_TOKEN_SIGNING_KEY"
	if os.Getenv(param) == "" {
		defaultVal := "gHs@dHk4Bs#v5HeK4h"
		cfg.HashingTokenSigningKey = defaultVal
		logger.Print(msg.WarningCannotConvertCvar(param, defaultVal))
	} else {
		cfg.HashingTokenSigningKey = os.Getenv(param)
	}

	param = "SERVER_HOST"
	if os.Getenv(param) == "" {
		defaultVal := "localhost"
		cfg.ServerHost = defaultVal
		logger.Print(msg.WarningCannotConvertCvar(param, defaultVal))
	} else {
		cfg.ServerHost = os.Getenv(param)
	}

	param = "SERVER_PORT"
	if val, err := strconv.Atoi(os.Getenv(param)); err != nil {
		defaultVal := 8000
		cfg.ServerPort = defaultVal
		logger.Print(msg.WarningCannotConvertCvar(param, defaultVal))
	} else {
		cfg.ServerPort = val
	}

	param = "SERVER_READ_TIMEOUT_SECONDS"
	if val, err := strconv.Atoi(os.Getenv(param)); err != nil {
		defaultVal := 15
		cfg.ServerReadTimeoutSeconds = defaultVal
		logger.Print(msg.WarningCannotConvertCvar(param, defaultVal))
	} else {
		cfg.ServerReadTimeoutSeconds = val
	}

	param = "SERVER_WRITE_TIMEOUT_SECONDS"
	if val, err := strconv.Atoi(os.Getenv(param)); err != nil {
		defaultVal := 15
		cfg.ServerWriteTimeoutSeconds = defaultVal
		logger.Print(msg.WarningCannotConvertCvar(param, defaultVal))
	} else {
		cfg.ServerWriteTimeoutSeconds = val
	}

	param = "pagination.getLimitDefault"
	if val := viper.GetInt(param); val == 0 {
		defaultVal := 30
		cfg.PaginationGetLimitDefault = defaultVal
		logger.Print(msg.WarningCannotConvertCvar(param, defaultVal))
	} else {
		cfg.PaginationGetLimitDefault = val
	}

	cfg.ServerDebugEnable = viper.GetBool("server.debugEnable")

	param = "server.maxHeaderBytes"
	if val := viper.GetInt(param); val == 0 {
		defaultVal := 1048576 // 1 megabyte
		cfg.ServerMaxHeaderBytes = defaultVal
		logger.Print(msg.WarningCannotConvertCvar(param, defaultVal))
	} else {
		cfg.ServerMaxHeaderBytes = val
	}

	param = "session.ttlHours"
	if val := viper.GetInt(param); val == 0 {
		defaultVal := 168 // 7 days
		cfg.SessionTTLHours = defaultVal
		logger.Print(msg.WarningCannotConvertCvar(param, defaultVal))
	} else {
		cfg.SessionTTLHours = val
	}

	param = "stream.iceServers"
	if val := viper.GetStringSlice(param); len(val) == 0 {
		defaultICEServer := "stun:stun.l.google.com:19302"
		defaultVal := []string{defaultICEServer}
		cfg.StreamICEServers = defaultVal
		logger.Print(msg.WarningCannotConvertCvar(param, "[\""+defaultICEServer+"\"]"))
	} else {
		cfg.StreamICEServers = val
	}

	param = "stream.snapshotPeriodSeconds"
	if val := viper.GetInt(param); val == 0 {
		defaultVal := 60
		cfg.StreamSnapshotPeriodSeconds = defaultVal
		logger.Print(msg.WarningCannotConvertCvar(param, defaultVal))
	} else {
		cfg.StreamSnapshotPeriodSeconds = val
	}

	cfg.StreamSnapshotShowStatus = viper.GetBool("stream.snapshotShowStatus")
	cfg.StreamSnapshotsEnable = viper.GetBool("stream.snapshotsEnable")

	param = "stream.streamsUpdatePeriodSeconds"
	if val := viper.GetInt(param); val == 0 {
		defaultVal := 60
		cfg.StreamStreamsUpdatePeriodSeconds = defaultVal
		logger.Print(msg.WarningCannotConvertCvar(param, defaultVal))
	} else {
		cfg.StreamStreamsUpdatePeriodSeconds = val
	}

	return &cfg, nil
}
